pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+	cccutil "k8s.io/kubernetes/pkg/controller/util/clustercidrconfig"
+	controllerutil "k8s.io/kubernetes/pkg/controller/util/node"
+	"k8s.io/kubernetes/pkg/util/slice"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultClusterCIDRConfigKey        = "kubernetes.io/clusterCIDRConfig"
+	defaultClusterCIDRConfigValue      = "default"
+	defaultClusterCIDRConfigName       = "default-cluster-cidr-config"
+	defaultClusterCIDRConfigAPIVersion = "networking.k8s.io/v1alpha1"
+	clusterCIDRConfigFinalizer         = "networking.k8s.io/cluster-cidr-config-finalizer"
+	ipv4MaxCIDRMask                    = 32
+	ipv6MaxCIDRMask                    = 128
+	minPerNodeHostBits                 = 4
+	maxRetries                         = 15
+)
+
+// CIDRs are reserved, then node resource is patched with them.
+// multiCIDRNodeReservedCIDRs holds the reservation info for a node.
+type multiCIDRNodeReservedCIDRs struct {
+	nodeReservedCIDRs
+	clusterCIDR *cidrset.ClusterCIDR
+}
+
+type multiCIDRRangeAllocator struct {
+	client clientset.Interface
+	// nodeLister is able to list/get nodes and is populated by the shared informer passed to controller.
+	nodeLister corelisters.NodeLister
+	// nodesSynced returns true if the node shared informer has been synced at least once.
+	nodesSynced cache.InformerSynced
+	// clusterCIDRConfigLister is able to list/get clustercidrconfigs and is populated by the shared informer passed to controller.
+	clusterCIDRConfigLister networkinglisters.ClusterCIDRConfigLister
+	// clusterCIDRConfigSynced returns true if the clustercidrconfig shared informer has been synced at least once.
+	clusterCIDRConfigSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
+	recorder              record.EventRecorder
+	// Keep a set of nodes that are currently being processed to avoid races in CIDR allocation.
+	lock              *sync.Mutex
+	nodesInProcessing sets.String
+	// queue is where incoming work is placed to de-dup and to allow "easy"
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+	// cidrMap maps ClusterCIDRConfig labels to ClusterCIDR objects.
+	cidrMap map[string][]*cidrset.ClusterCIDR
+}
+
+// NewMultiCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one for each ip family).
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	clusterCIDRConfigInformer networkinginformers.ClusterCIDRConfigInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList,
+	cccList *networkingv1alpha1.ClusterCIDRConfigList) (CIDRAllocator, error) {
```

Comment:
please do this so the function sig is clear in the code

```
func foo(
	x int,
	y bool, // <<< trailing comma,
) { // <<< ) { closing is indented to the left.
	fmt.Println("asdf")
}
```

Reply:
Done